src/v/cloud_storage/manifest.h

```diff
@@ -134,9 +139,10 @@ class manifest final : public base_manifest {
         auto operator<=>(const segment_meta&) const = default;
     };
 
-    using key = segment_name;
+    /// Segment key in the maifest
+    using key = segment_name_components;
     using value = segment_meta;
-    using segment_map = std::map<key, value>;
+    using segment_map = absl::btree_map<key, value>;
```

Comment:
is this where we lose pointer stability (presumably because the btree will split its internal nodes)?

> This one looks okay - I haven't found any places where we store iterators or pointers to map items.

Did you also consider beyond "store iterators or pointers" where pointers and iterators are held across scheduling points like co_await? Also pointer stability crops up in different ways that iterator invalidations. Generally the iterator invalidation rules are still relevant when switching to a new container, but pointers / references don't.

A tricky one is how things work when moving values (e.g. if they are internally moved in a rebalance action).

Reply:
I'm not familiar enough with the code base to know how much surface area needs to be checked, but if there is some doubt, you could store std::unique_ptr<value> as an interim step.

Reply:
I checked for scheduling points that can trigger manifest iterator invalidation but wasn't able to do so. It should be OK.